File: PhoneticTokenizer/go/phonetok/phones.go
// Package phonetok is a phonetic tokenizer: text as the sounds it is made of.
//
// It is the Go port of the Python package of the same name (../../phonetok),
// module for module, and reads the same two data files - the core lexicon and
// the letter-to-sound rules - so that the three ports (Python, Go, Rust) turn
// the same text into the same sounds, token for token.
//
// This file is the sound inventory: the 39 phonemes of the CMU Pronouncing
// Dictionary's ARPAbet, the stress digit a vowel carries, the fixed alphabet
// of ids, the articulatory features, the sonority scale and the IPA.
package phonetok

import (
	"fmt"
	"math"
	"strings"
)

// Vowels are the 15 vowel phonemes, in ARPAbet order.
var Vowels = []string{"AA", "AE", "AH", "AO", "AW", "AY", "EH", "ER", "EY", "IH", "IY", "OW", "OY", "UH", "UW"}

// Consonants are the 24 consonant phonemes, in ARPAbet order.
var Consonants = []string{"B", "CH", "D", "DH", "F", "G", "HH", "JH", "K", "L", "M", "N", "NG", "P", "R", "S", "SH", "T", "TH", "V", "W", "Y", "Z", "ZH"}

// The token grammar's fixed symbols.
const (
	Boundary      = "#"
	PauseShort    = ","
	PauseFull     = "."
	PauseQuestion = "?"
	PAD           = "<pad>"
	UNK           = "<unk>"
	BOS           = "<s>"
	EOS           = "</s>"
)

// Pauses are what punctuation becomes; Specials the four every tokenizer has.
var (
	Pauses   = []string{PauseShort, PauseFull, PauseQuestion}
	Specials = []string{PAD, UNK, BOS, EOS}
)

// Ids of the head of the alphabet.
const (
	PadID      = 0
	UnkID      = 1
	BosID      = 2
	EosID      = 3
	BoundaryID = 4
)

// Symbols is the fixed alphabet in id order: the specials, the boundary, the
// pauses, then every ARPAbet symbol as cmudict.symbols lists them (a bare vowel,
// its three stresses, then the consonants): 92 symbols.
var Symbols = func() []string {
	out := append([]string{}, Specials...)
	out = append(out, Boundary)
	out = append(out, Pauses...)
	for _, v := range Vowels {
		out = append(out, v, v+"0", v+"1", v+"2")
	}
	out = append(out, Consonants...)
	return out
}()

// SymbolID is symbol -> id.
var SymbolID = func() map[string]int {
	m := make(map[string]int, len(Symbols))
	for i, s := range Symbols {
		m[s] = i
	}
	return m
}()

var (
	vowelSet     = toSet(Vowels)
	consonantSet = toSet(Consonants)
)

func toSet(items []string) map[string]bool {
	m := make(map[string]bool, len(items))
	for _, s := range items {
		m[s] = true
	}
	return m
}

// Base is the phoneme without its stress digit: "AH0" -> "AH", "K" -> "K".
func Base(phone string) string {
	if n := len(phone); n > 0 && phone[n-1] >= '0' && phone[n-1] <= '2' {
		return phone[:n-1]
	}
	return phone
}

// StressOf is the stress digit of a vowel, or -1 for a consonant or a bare vowel.
func StressOf(phone string) int {
	if n := len(phone); n > 0 && phone[n-1] >= '0' && phone[n-1] <= '2' {
		return int(phone[n-1] - '0')
	}
	return -1
}

// IsVowel reports whether this is a vowel, with or without its stress digit.
func IsVowel(phone string) bool { return vowelSet[Base(phone)] }

// IsConsonant reports whether this is a consonant.
func IsConsonant(phone string) bool { return consonantSet[phone] }

// IsPhone reports whether token is an ARPAbet symbol: a consonant, or a vowel with or without a stress digit.
func IsPhone(token string) bool { return vowelSet[Base(token)] || consonantSet[token] }

// IsStressed is primary or secondary stress.
func IsStressed(phone string) bool { s := StressOf(phone); return s == 1 || s == 2 }

// StripStress is the same phones with every stress digit removed.
func StripStress(phones []string) []string {
	out := make([]string, len(phones))
	for i, p := range phones {
		out[i] = Base(p)
	}
	return out
}

// WithStress is a vowel with its stress set; a consonant is left alone.
func WithStress(phone string, stress int) string {
	b := Base(phone)
	if vowelSet[b] {
		return fmt.Sprintf("%s%d", b, stress)
	}
	return phone
}

// CheckPhones returns an error at the first token that is not a phone.
func CheckPhones(phones []string) error {
	for _, p := range phones {
		if !IsPhone(p) {
			return fmt.Errorf("%q is not an ARPAbet phone", p)
		}
	}
	return nil
}

// -- classes and articulation ----------------------------------------------------

var (
	Stops            = toSet([]string{"B", "D", "G", "K", "P", "T"})
	Affricates       = toSet([]string{"CH", "JH"})
	Fricatives       = toSet([]string{"DH", "F", "S", "SH", "TH", "V", "Z", "ZH"})
	Aspirates        = toSet([]string{"HH"})
	Nasals           = toSet([]string{"M", "N", "NG"})
	Liquids          = toSet([]string{"L", "R"})
	Glides           = toSet([]string{"W", "Y"})
	Sibilants        = toSet([]string{"S", "Z", "SH", "ZH", "CH", "JH"})
	VoicedConsonants = toSet([]string{"B", "D", "DH", "G", "JH", "L", "M", "N", "NG", "R", "V", "W", "Y", "Z", "ZH"})
	labial           = toSet([]string{"B", "P", "M", "F", "V", "W"})
	coronal          = toSet([]string{"T", "D", "S", "Z", "N", "L", "R", "TH", "DH", "SH", "ZH", "CH", "JH"})
	dorsal           = toSet([]string{"K", "G", "NG", "W", "Y"})
	glottal          = toSet([]string{"HH"})
	strident         = toSet([]string{"S", "Z", "SH", "ZH", "CH", "JH", "F", "V"})
	Diphthongs       = toSet([]string{"AW", "AY", "EY", "OW", "OY"})
	RhoticVowels     = toSet([]string{"ER"})
	continuant       = func() map[string]bool {
		m := map[string]bool{}
		for _, set := range []map[string]bool{Fricatives, Aspirates, Liquids, Glides} {
			for k := range set {
				m[k] = true
			}
		}
		return m
	}()
)

// IsSonorant reports whether a consonant is a nasal, a liquid or a glide.
func IsSonorant(c string) bool { return Nasals[c] || Liquids[c] || Glides[c] }

// Sonority is where a phone sits on the scale, 0 (a voiceless stop) to 10 (a low vowel); the stress digit is ignored.
func Sonority(phone string) int {
	b := Base(phone)
	switch {
	case Stops[b]:
		if VoicedConsonants[b] {
			return 1
		}
		return 0
	case Affricates[b]:
		if VoicedConsonants[b] {
			return 2
		}
		return 1
	case Fricatives[b]:
		if VoicedConsonants[b] {
			return 3
		}
		return 2
	case b == "HH":
		return 2
	case Nasals[b]:
		return 4
	case b == "L":
		return 5
	case b == "R":
		return 6
	case Glides[b]:
		return 7
	}
	switch b {
	case "IY", "IH", "UW", "UH", "ER":
		return 8
	case "EY", "EH", "OW", "AO", "AH", "OY":
		return 9
	case "AE", "AA", "AY", "AW":
		return 10
	}
	panic("not a phone: " + phone)
}

// MaxSonority is the top of the scale.
const MaxSonority = 10

type vowelSpace struct {
	height, back, round, tense float64
	glide                      string
}

var vowelSpaces = map[string]vowelSpace{
	"IY": {1.0, 0.0, 0.0, 1.0, ""}, "IH": {0.8, 0.2, 0.0, 0.0, ""}, "EY": {0.6, 0.0, 0.0, 1.0, "Y"},
	"EH": {0.5, 0.2, 0.0, 0.0, ""}, "AE": {0.2, 0.2, 0.0, 0.0, ""}, "AA": {0.0, 0.9, 0.0, 1.0, ""},
	"AO": {0.3, 1.0, 1.0, 1.0, ""}, "OW": {0.6, 1.0, 1.0, 1.0, "W"}, "UH": {0.8, 0.9, 1.0, 0.0, ""},
	"UW": {1.0, 1.0, 1.0, 1.0, ""}, "AH": {0.4, 0.6, 0.0, 0.0, ""}, "ER": {0.5, 0.5, 0.0, 1.0, ""},
	"AY": {0.1, 0.5, 0.0, 1.0, "Y"}, "AW": {0.1, 0.5, 0.0, 1.0, "W"}, "OY": {0.3, 1.0, 1.0, 1.0, "Y"},
}

// FeatureNames are the 22 dimensions of Features, in order.
var FeatureNames = []string{
	"syllabic", "consonantal", "sonorant", "voice", "nasal", "continuant", "strident", "lateral", "rhotic",
	"labial", "coronal", "dorsal", "glottal",
	"high", "low", "back", "round", "tense", "glide_y", "glide_w",
	"stress", "sonority",
}

// FeatureDim is the length of a feature vector.
const FeatureDim = 22

func b2f(b bool) float64 {
	if b {
		return 1
	}
	return 0
}

// Features is the articulatory feature vector of a phone (FeatureNames), stress included.
func Features(phone string) ([]float64, error) {
	b := Base(phone)
	if !vowelSet[b] && !consonantSet[b] {
		return nil, fmt.Errorf("%q is not an ARPAbet phone", phone)
	}
	stress := 0.0
	switch StressOf(phone) {
	case 1:
		stress = 1.0
	case 2:
		stress = 0.5
	}
	son := float64(Sonority(b)) / MaxSonority
	if vowelSet[b] {
		v := vowelSpaces[b]
		return []float64{
			1, 0, 1, 1, 0, 1, 0, 0, b2f(RhoticVowels[b]),
			0, 0, 0, 0,
			v.height, 1 - v.height, v.back, v.round, v.tense, b2f(v.glide == "Y"), b2f(v.glide == "W"),
			stress, son,
		}, nil
	}
	return []float64{
		0,
		b2f(!(Glides[b] || Aspirates[b])),
		b2f(IsSonorant(b)),
		b2f(VoicedConsonants[b]),
		b2f(Nasals[b]),
		b2f(continuant[b]),
		b2f(strident[b]),
		b2f(b == "L"),
		b2f(b == "R"),
		b2f(labial[b]),
		b2f(coronal[b]),
		b2f(dorsal[b]),
		b2f(glottal[b]),
		b2f(b == "Y" || b == "W"),
		0,
		b2f(b == "W" || b == "K" || b == "G" || b == "NG"),
		b2f(b == "W"),
		0, 0, 0,
		0, son,
	}, nil
}

// Distance is how differently two phones are made: the Euclidean distance of their feature vectors.
func Distance(a, b string) float64 {
	fa, ea := Features(a)
	fb, eb := Features(b)
	if ea != nil || eb != nil {
		panic("not a phone")
	}
	sum := 0.0
	for i := range fa {
		d := fa[i] - fb[i]
		sum += d * d
	}
	return math.Sqrt(sum)
}

// Similarity is 1 for the same phone, falling towards 0.
func Similarity(a, b string) float64 { return 1 / (1 + Distance(a, b)) }

// -- the IPA ---------------------------------------------------------------------

// IPA is ARPAbet -> IPA, one entry per phoneme.
var IPA = map[string]string{
	"AA": "ɑ", "AE": "æ", "AH": "ʌ", "AO": "ɔ", "AW": "aʊ", "AY": "aɪ", "EH": "ɛ", "ER": "ɝ", "EY": "eɪ",
	"IH": "ɪ", "IY": "i", "OW": "oʊ", "OY": "ɔɪ", "UH": "ʊ", "UW": "u",
	"B": "b", "CH": "tʃ", "D": "d", "DH": "ð", "F": "f", "G": "ɡ", "HH": "h", "JH": "dʒ", "K": "k", "L": "l",
	"M": "m", "N": "n", "NG": "ŋ", "P": "p", "R": "ɹ", "S": "s", "SH": "ʃ", "T": "t", "TH": "θ", "V": "v",
	"W": "w", "Y": "j", "Z": "z", "ZH": "ʒ",
}

var ipaReduced = map[string]string{"AH": "ə", "ER": "ɚ"}

// ToIPA is phones as one IPA string, a stress mark before a stressed vowel (when stress is true).
func ToIPA(phones []string, stress bool) string {
	var b strings.Builder
	for _, p := range phones {
		base := Base(p)
		s := StressOf(p)
		switch {
		case vowelSet[base]:
			if stress && s == 1 {
				b.WriteString("ˈ")
			} else if stress && s == 2 {
				b.WriteString("ˌ")
			}
			if r, ok := ipaReduced[base]; ok && s == 0 {
				b.WriteString(r)
			} else {
				b.WriteString(IPA[base])
			}
		case consonantSet[base]:
			b.WriteString(IPA[base])
		default:
			b.WriteString(p)
		}
	}
	return b.String()
}

// -- the possessive / plural / past suffixes ---------------------------------------

// PluralSuffix is the -s / -'s ending after these phones: IH0 Z after a sibilant, Z after voice, S otherwise.
func PluralSuffix(phones []string) []string {
	if len(phones) == 0 {
		return []string{"Z"}
	}
	last := Base(phones[len(phones)-1])
	switch {
	case Sibilants[last]:
		return []string{"IH0", "Z"}
	case vowelSet[last] || VoicedConsonants[last]:
		return []string{"Z"}
	}
	return []string{"S"}
}

// PastSuffix is the -ed ending: IH0 D after T or D, D after voice, T otherwise.
func PastSuffix(phones []string) []string {
	if len(phones) == 0 {
		return []string{"D"}
	}
	last := Base(phones[len(phones)-1])
	switch {
	case last == "T" || last == "D":
		return []string{"IH0", "D"}
	case vowelSet[last] || VoicedConsonants[last]:
		return []string{"D"}
	}
	return []string{"T"}
}

File: PhoneticTokenizer/rust/src/phones.rs
//! The sound inventory (the port of phones.py): the 39 ARPAbet phonemes, the
//! stress digit a vowel carries, the fixed alphabet of ids, the articulatory
//! features, the sonority scale and the IPA.

use std::collections::HashMap;
use std::sync::OnceLock;

/// The 15 vowel phonemes, in ARPAbet order.
pub const VOWELS: [&str; 15] = [
    "AA", "AE", "AH", "AO", "AW", "AY", "EH", "ER", "EY", "IH", "IY", "OW", "OY", "UH", "UW",
];

/// The 24 consonant phonemes, in ARPAbet order.
pub const CONSONANTS: [&str; 24] = [
    "B", "CH", "D", "DH", "F", "G", "HH", "JH", "K", "L", "M", "N", "NG", "P", "R", "S", "SH", "T",
    "TH", "V", "W", "Y", "Z", "ZH",
];

pub const BOUNDARY: &str = "#";
pub const PAUSE_SHORT: &str = ",";
pub const PAUSE_FULL: &str = ".";
pub const PAUSE_QUESTION: &str = "?";
pub const PAUSES: [&str; 3] = [PAUSE_SHORT, PAUSE_FULL, PAUSE_QUESTION];
pub const PAD: &str = "<pad>";
pub const UNK: &str = "<unk>";
pub const BOS: &str = "<s>";
pub const EOS: &str = "</s>";
pub const SPECIALS: [&str; 4] = [PAD, UNK, BOS, EOS];

pub const PAD_ID: usize = 0;
pub const UNK_ID: usize = 1;
pub const BOS_ID: usize = 2;
pub const EOS_ID: usize = 3;
pub const BOUNDARY_ID: usize = 4;

/// The fixed alphabet in id order: the specials, the boundary, the pauses, then every ARPAbet
/// symbol as cmudict.symbols lists them: 92 symbols.
pub fn symbols() -> &'static Vec<String> {
    static SYMBOLS: OnceLock<Vec<String>> = OnceLock::new();
    SYMBOLS.get_or_init(|| {
        let mut out: Vec<String> = SPECIALS.iter().map(|s| s.to_string()).collect();
        out.push(BOUNDARY.to_string());
        out.extend(PAUSES.iter().map(|s| s.to_string()));
        for v in VOWELS {
            out.push(v.to_string());
            for d in ["0", "1", "2"] {
                out.push(format!("{v}{d}"));
            }
        }
        out.extend(CONSONANTS.iter().map(|s| s.to_string()));
        out
    })
}

/// Symbol -> id.
pub fn symbol_id(symbol: &str) -> Option<usize> {
    static IDS: OnceLock<HashMap<String, usize>> = OnceLock::new();
    IDS.get_or_init(|| {
        symbols()
            .iter()
            .enumerate()
            .map(|(i, s)| (s.clone(), i))
            .collect()
    })
    .get(symbol)
    .copied()
}

/// The phoneme without its stress digit.
pub fn base(phone: &str) -> &str {
    match phone.as_bytes().last() {
        Some(b'0'..=b'2') => &phone[..phone.len() - 1],
        _ => phone,
    }
}

/// The stress digit of a vowel, or `None` for a consonant or a bare vowel.
pub fn stress_of(phone: &str) -> Option<u8> {
    match phone.as_bytes().last() {
        Some(d @ b'0'..=b'2') => Some(d - b'0'),
        _ => None,
    }
}

pub fn is_vowel(phone: &str) -> bool {
    VOWELS.contains(&base(phone))
}

pub fn is_consonant(phone: &str) -> bool {
    CONSONANTS.contains(&phone)
}

/// Is this an ARPAbet symbol: a consonant, or a vowel with or without a stress digit?
pub fn is_phone(token: &str) -> bool {
    is_vowel(token) || is_consonant(token)
}

pub fn is_stressed(phone: &str) -> bool {
    matches!(stress_of(phone), Some(1) | Some(2))
}

pub fn strip_stress(phones: &[String]) -> Vec<String> {
    phones.iter().map(|p| base(p).to_string()).collect()
}

/// A vowel with its stress set; a consonant is left alone.
pub fn with_stress(phone: &str, stress: u8) -> String {
    let b = base(phone);
    if VOWELS.contains(&b) {
        format!("{b}{stress}")
    } else {
        phone.to_string()
    }
}

pub fn check_phones(phones: &[String]) -> Result<(), String> {
    for p in phones {
        if !is_phone(p) {
            return Err(format!("{p:?} is not an ARPAbet phone"));
        }
    }
    Ok(())
}

// -- classes and articulation ----------------------------------------------------

pub const STOPS: [&str; 6] = ["B", "D", "G", "K", "P", "T"];
pub const AFFRICATES: [&str; 2] = ["CH", "JH"];
pub const FRICATIVES: [&str; 8] = ["DH", "F", "S", "SH", "TH", "V", "Z", "ZH"];
pub const NASALS: [&str; 3] = ["M", "N", "NG"];
pub const LIQUIDS: [&str; 2] = ["L", "R"];
pub const GLIDES: [&str; 2] = ["W", "Y"];
pub const SIBILANTS: [&str; 6] = ["S", "Z", "SH", "ZH", "CH", "JH"];
pub const VOICED_CONSONANTS: [&str; 15] = [
    "B", "D", "DH", "G", "JH", "L", "M", "N", "NG", "R", "V", "W", "Y", "Z", "ZH",
];
const LABIAL: [&str; 6] = ["B", "P", "M", "F", "V", "W"];
const CORONAL: [&str; 13] = [
    "T", "D", "S", "Z", "N", "L", "R", "TH", "DH", "SH", "ZH", "CH", "JH",
];
const DORSAL: [&str; 5] = ["K", "G", "NG", "W", "Y"];
const STRIDENT: [&str; 8] = ["S", "Z", "SH", "ZH", "CH", "JH", "F", "V"];

pub fn is_sibilant(c: &str) -> bool {
    SIBILANTS.contains(&c)
}

pub fn is_voiced_consonant(c: &str) -> bool {
    VOICED_CONSONANTS.contains(&c)
}

pub fn is_sonorant(c: &str) -> bool {
    NASALS.contains(&c) || LIQUIDS.contains(&c) || GLIDES.contains(&c)
}

pub const MAX_SONORITY: u32 = 10;

/// Where a phone sits on the sonority scale, 0 (a voiceless stop) to 10 (a low vowel).
pub fn sonority(phone: &str) -> u32 {
    let b = base(phone);
    let voiced = VOICED_CONSONANTS.contains(&b);
    if STOPS.contains(&b) {
        return if voiced { 1 } else { 0 };
    }
    if AFFRICATES.contains(&b) {
        return if voiced { 2 } else { 1 };
    }
    if FRICATIVES.contains(&b) {
        return if voiced { 3 } else { 2 };
    }
    match b {
        "HH" => 2,
        "M" | "N" | "NG" => 4,
        "L" => 5,
        "R" => 6,
        "W" | "Y" => 7,
        "IY" | "IH" | "UW" | "UH" | "ER" => 8,
        "EY" | "EH" | "OW" | "AO" | "AH" | "OY" => 9,
        "AE" | "AA" | "AY" | "AW" => 10,
        _ => panic!("not a phone: {phone}"),
    }
}

/// The 22 dimensions of `features`, in order.
pub const FEATURE_NAMES: [&str; 22] = [
    "syllabic",
    "consonantal",
    "sonorant",
    "voice",
    "nasal",
    "continuant",
    "strident",
    "lateral",
    "rhotic",
    "labial",
    "coronal",
    "dorsal",
    "glottal",
    "high",
    "low",
    "back",
    "round",
    "tense",
    "glide_y",
    "glide_w",
    "stress",
    "sonority",
];
pub const FEATURE_DIM: usize = 22;

fn vowel_space(v: &str) -> (f64, f64, f64, f64, &'static str) {
    match v {
        "IY" => (1.0, 0.0, 0.0, 1.0, ""),
        "IH" => (0.8, 0.2, 0.0, 0.0, ""),
        "EY" => (0.6, 0.0, 0.0, 1.0, "Y"),
        "EH" => (0.5, 0.2, 0.0, 0.0, ""),
        "AE" => (0.2, 0.2, 0.0, 0.0, ""),
        "AA" => (0.0, 0.9, 0.0, 1.0, ""),
        "AO" => (0.3, 1.0, 1.0, 1.0, ""),
        "OW" => (0.6, 1.0, 1.0, 1.0, "W"),
        "UH" => (0.8, 0.9, 1.0, 0.0, ""),
        "UW" => (1.0, 1.0, 1.0, 1.0, ""),
        "AH" => (0.4, 0.6, 0.0, 0.0, ""),
        "ER" => (0.5, 0.5, 0.0, 1.0, ""),
        "AY" => (0.1, 0.5, 0.0, 1.0, "Y"),
        "AW" => (0.1, 0.5, 0.0, 1.0, "W"),
        "OY" => (0.3, 1.0, 1.0, 1.0, "Y"),
        _ => unreachable!(),
    }
}

fn b2f(b: bool) -> f64 {
    if b {
        1.0
    } else {
        0.0
    }
}

/// The articulatory feature vector of a phone, stress included.
pub fn features(phone: &str) -> Result<[f64; FEATURE_DIM], String> {
    let b = base(phone);
    if !is_vowel(b) && !is_consonant(b) {
        return Err(format!("{phone:?} is not an ARPAbet phone"));
    }
    let stress = match stress_of(phone) {
        Some(1) => 1.0,
        Some(2) => 0.5,
        _ => 0.0,
    };
    let son = sonority(b) as f64 / MAX_SONORITY as f64;
    if is_vowel(b) {
        let (height, back, round, tense, glide) = vowel_space(b);
        return Ok([
            1.0,
            0.0,
            1.0,
            1.0,
            0.0,
            1.0,
            0.0,
            0.0,
            b2f(b == "ER"),
            0.0,
            0.0,
            0.0,
            0.0,
            height,
            1.0 - height,
            back,
            round,
            tense,
            b2f(glide == "Y"),
            b2f(glide == "W"),
            stress,
            son,
        ]);
    }
    let continuant =
        FRICATIVES.contains(&b) || b == "HH" || LIQUIDS.contains(&b) || GLIDES.contains(&b);
    Ok([
        0.0,
        b2f(!(GLIDES.contains(&b) || b == "HH")),
        b2f(is_sonorant(b)),
        b2f(VOICED_CONSONANTS.contains(&b)),
        b2f(NASALS.contains(&b)),
        b2f(continuant),
        b2f(STRIDENT.contains(&b)),
        b2f(b == "L"),
        b2f(b == "R"),
        b2f(LABIAL.contains(&b)),
        b2f(CORONAL.contains(&b)),
        b2f(DORSAL.contains(&b)),
        b2f(b == "HH"),
        b2f(b == "Y" || b == "W"),
        0.0,
        b2f(matches!(b, "W" | "K" | "G" | "NG")),
        b2f(b == "W"),
        0.0,
        0.0,
        0.0,
        0.0,
        son,
    ])
}

/// How differently two phones are made: the Euclidean distance of their feature vectors.
pub fn distance(a: &str, b: &str) -> f64 {
    let (fa, fb) = (features(a).expect("a phone"), features(b).expect("a phone"));
    fa.iter()
        .zip(fb.iter())
        .map(|(x, y)| (x - y) * (x - y))
        .sum::<f64>()
        .sqrt()
}

pub fn similarity(a: &str, b: &str) -> f64 {
    1.0 / (1.0 + distance(a, b))
}

// -- the IPA ---------------------------------------------------------------------

/// ARPAbet -> IPA, one entry per phoneme.
pub fn ipa(phoneme: &str) -> &'static str {
    match phoneme {
        "AA" => "ɑ",
        "AE" => "æ",
        "AH" => "ʌ",
        "AO" => "ɔ",
        "AW" => "aʊ",
        "AY" => "aɪ",
        "EH" => "ɛ",
        "ER" => "ɝ",
        "EY" => "eɪ",
        "IH" => "ɪ",
        "IY" => "i",
        "OW" => "oʊ",
        "OY" => "ɔɪ",
        "UH" => "ʊ",
        "UW" => "u",
        "B" => "b",
        "CH" => "tʃ",
        "D" => "d",
        "DH" => "ð",
        "F" => "f",
        "G" => "ɡ",
        "HH" => "h",
        "JH" => "dʒ",
        "K" => "k",
        "L" => "l",
        "M" => "m",
        "N" => "n",
        "NG" => "ŋ",
        "P" => "p",
        "R" => "ɹ",
        "S" => "s",
        "SH" => "ʃ",
        "T" => "t",
        "TH" => "θ",
        "V" => "v",
        "W" => "w",
        "Y" => "j",
        "Z" => "z",
        "ZH" => "ʒ",
        _ => "",
    }
}

/// Phones as one IPA string, a stress mark before a stressed vowel when `stress` is true.
pub fn to_ipa(phones: &[String], stress: bool) -> String {
    let mut out = String::new();
    for p in phones {
        let b = base(p);
        let s = stress_of(p);
        if is_vowel(b) {
            if stress && s == Some(1) {
                out.push('ˈ');
            } else if stress && s == Some(2) {
                out.push('ˌ');
            }
            match (b, s) {
                ("AH", Some(0)) => out.push('ə'),
                ("ER", Some(0)) => out.push('ɚ'),
                _ => out.push_str(ipa(b)),
            }
        } else if is_consonant(b) {
            out.push_str(ipa(b));
        } else {
            out.push_str(p);
        }
    }
    out
}

// -- the possessive / plural / past suffixes --------------------------------------

/// The -s / -'s ending after these phones.
pub fn plural_suffix(phones: &[String]) -> Vec<String> {
    let Some(last) = phones.last() else {
        return vec!["Z".into()];
    };
    let last = base(last);
    if is_sibilant(last) {
        vec!["IH0".into(), "Z".into()]
    } else if is_vowel(last) || is_voiced_consonant(last) {
        vec!["Z".into()]
    } else {
        vec!["S".into()]
    }
}

/// The -ed ending.
pub fn past_suffix(phones: &[String]) -> Vec<String> {
    let Some(last) = phones.last() else {
        return vec!["D".into()];
    };
    let last = base(last);
    if last == "T" || last == "D" {
        vec!["IH0".into(), "D".into()]
    } else if is_vowel(last) || is_voiced_consonant(last) {
        vec!["D".into()]
    } else {
        vec!["T".into()]
    }
}

/// A helper used across the crate: owned strings from string slices.
pub fn strings(items: &[&str]) -> Vec<String> {
    items.iter().map(|s| s.to_string()).collect()
}
